README.md
# Special Cooked Collections

An extension of Collection package in Dart. It provides useful methods to work with collections
such as: Tuple, Stack and SpecialQueue.

## Features

You can create Tuple, Stack and SpecialQueue objects that you can manipulate with
the help of their methods.

- Tuple: A Tuple is a collection with fixed length and elements.
- Stack: A Stack is a collection that can be manipulated at one end.
- SpecialQueue: A SpecialQueue is a collection that can be manipulated at both ends.

## Installing

#### Depend on it

Run this command:

    $ dart pub add collection_extension

This will add a line like this to your package's pubspec.yaml (and run an implicit dart pub get):

    dependencies:
        shelf: ^1.3.2

#### Import it

Now in your Dart code, you can use:

    import 'package:collection_extension/collection_extension.dart'

## Usage

The next snippet is a code example using the library.

```dart
final tuple = Tuple.fromList([5, 9, 11, 3, 15, 19]);
  print(tuple.contains(5)); // true
  print(tuple.contains('5')); //false
  print(tuple
      .firstWhere((element) => element is num && element > 10 && element < 20)); // 11

final stack = Stack();
  stack.push(3); // [3]
  stack.push(9); // [3, 9]
  stack.push('a'); // [3, 9, a]
  stack.popAll(); // []
  stack.pushAll([25, 35, 99, 199, 399, 9999999]); // [25, 35, 99, 199, 399, 9999999]
  print(stack.canPop); // true

final specialQueue = SpecialQueue();
   specialQueue.push(5); // [5]
   specialQueue.push(33); // [5, 33]
   specialQueue.removeFirst(); // [33]
   specialQueue.push(44); // [33, 44]
   print(specialQueue.peak); // [33]
   specialQueue.pushAll([20, 50, 34, 127]); // [33, 44, 20, 50, 34, 127]
```

## Additional information

This is just an experimental package, it's usage is up to the developer.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2022-09-06

### Added

- Initial version.

[Unreleased]: https://github.com/EmanuelLamba/collection_extension
[1.0.0]: https://github.com/EmanuelLamba/collection_extension


lib/src/stack.dart
/// A [Stack] is a `collection` that can be manipulated at one end.
/// One can execute the stack using [executeStack] method.
///
/// The [Stack] can be either created empty or with `elements` given
/// through a [List].
///
/// For simplicity reasons [Stack] is a [List] that can manipulate elements
/// only at one end.
///
/// Example:
/// ```dart
/// final stack = Stack();
/// print(stack.runtimeType); // Stack<dynamic>
///
/// //Adding elements to the stack
/// stack.push(5);
/// stack.pushAll([22, 'b', 9]);
/// print(stack); // [5, 22, 'b', 9];
///
/// // Removing elements from the stack
/// // (Elements can be removed only from the end of the stack)
/// stack.pop();
/// print(stack); // [5, 22, 'b']
/// ```
class Stack<E> {
  /// The [List] used to store the [Stack] and its `elements`.
  final List _stackList;

  /// Whether this `collection` has no elements.
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack();
  /// print(stack.isEmpty); // true
  /// ```
  bool get isEmpty => _stackList.isEmpty;

  /// Whether this `collection` has elements to pop.
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack();
  /// print(stack.canPop); // false
  /// ```
  bool get canPop => _stackList.isNotEmpty;

  /// The number of `elements` in this `collection`.
  ///
  /// The valid indices are `0` through `length - 1`
  ///
  /// ```dart
  /// final stack = Stack.fromList([2, 5, 9, 44]);
  /// print(stack.length); // 4
  /// ```
  int get length => _stackList.length;

  /// Creates an empty [Stack] object.
  ///
  /// The `constructor` will initialise the [List] that manages
  /// the [Stack] object.
  ///
  /// ```dart
  /// final stack = Stack();
  /// print(stack); // []
  /// ```
  Stack() : _stackList = <E>[];

  /// Creates a [Stack] object containing the elements of the
  /// [List] passed as argument.
  ///
  /// The [List] that mananges [Stack] will be initialised with
  /// the elements of the [List] passed as argument.
  ///
  /// ```dart
  /// final stack = Stack.fromList([2, 4, 8, 'a']);
  /// print(stack); // [2, 4, 8, 'a']
  /// ```
  Stack.fromList(List<E> list) : _stackList = list;

  /// Inserts [element] to the end of the [Stack].
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack.fromList([7, 8]);
  /// stack.push(5);
  /// print(stack); // [7, 8, 5]
  /// ```
  void push(E element) => _stackList.add(element);

  /// Inserts all elements from [iterable] to the end of [Stack].
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack.fromList([2, 4]);
  /// stack.pushAll([5, 9, 77, 'a']);
  /// print(stack); // [2, 4, 5, 9, 77, 'a']
  /// ```
  void pushAll(Iterable<E> iterable) => _stackList.addAll(iterable);

  /// Removes the last `element` of the [Stack].
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack.fromList([5, 9, 'a']);
  /// stack.pop();
  /// print(stack); // [5, 9]
  /// ```
  void pop() {
    if (canPop) _stackList.removeLast();
  }

  /// Checks whether [Stack] contains [element] or not.
  ///
  /// A `bool` object will be returned.
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack.fromList([5, 9, 'a', 'b']);
  /// bool hasElement = stack.contains('a');
  /// print(hasElement); // true
  /// ```
  bool contains(E element) {
    for (E item in _stackList) {
      if (item == element) return true;
    }
    return false;
  }

  /// The [Stack] will be executed.
  ///
  /// Every `element` will be printed in order and removed from
  /// the [Stack].
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack.fromList([5, 9, 'a']);
  /// stack.executeQueue(); // a 9 5
  /// print(stack); // []
  /// ```
  void executeStack() {
    for (E element in _stackList.reversed) {
      print(element);
    }
    popAll();
  }

  /// Removes all `elements` of the [Stack].
  ///
  /// Example:
  /// ```dart
  /// final stack = Stack.fromList([5, 9, 'a']);
  /// stack.popAll();
  /// print(stack); // []
  /// ```
  void popAll() {
    if (canPop) _stackList.removeRange(0, _stackList.length);
  }

  @override
  String toString() {
    return _stackList.toString();
  }
}


lib/src/tuple.dart
/// A [Tuple] is a `collection` with fixed length and elements.
///
/// The [Tuple] is created by initialising the [List] that manages
/// the [Tuple] with the elements of [List] passed as argument to
/// the constructor.
///
/// It can be said that a [Tuple] is a [List] with fixed `length` and `elements`.
///
/// Example:
/// ```dart
/// final tuple = Tuple.fromList([5, 7, 15, 20, 'a', 'b']);
/// print(tuple.runtimeType); // Tuple<Object>
///
/// // Checking for elements in tuple
/// tuple.contains(15); // true
/// print(tuple.firstWhere((element) => element is num && element > 10))); // 15
/// ```
class Tuple<E> {
  /// The [List] used to store the [Tuple] and its `elements`.
  final List _tupleList;

  /// Whether this `collection` has no elements.
  ///
  /// Example:
  /// ```dart
  /// final tuple = Tuple();
  /// print(tuple.isEmpty); // true
  /// ```
  bool get isEmpty => _tupleList.isEmpty;

  /// The number of `elements` in this `collection`.
  ///
  /// The valid indices are `0` through `length - 1`
  ///
  /// ```dart
  /// final tuple = Tuple.fromList([2, 5, 9, 44]);
  /// print(tuple.length); // 4
  /// ```
  int get length => _tupleList.length;

  /// Creates a [Tuple] object containing the elements of
  /// the [List] passed as argument.
  ///
  /// The [List] that manages [Tuple] will be initialised with
  /// the elements of the [List] passed as argument.
  ///
  /// ```dart
  /// final tuple = Tuple.fromList([5, 9, 30, 'w']);
  /// print(tuple); // [5, 9, 30, w]
  /// ```
  Tuple.fromList(List<E> list) : _tupleList = list;

  /// Checks whether [Tuple] contains [element] or not.
  ///
  /// A `bool` object will be returned.
  ///
  /// Example:
  /// ```dart
  /// final tuple = Tuple.fromList([5, 9, 'a', 'b']);
  /// bool hasElement = tuple.contains('a');
  /// print(hasElement); // true
  /// ```
  bool contains(E element) {
    for (E item in _tupleList) {
      if (item == element) return true;
    }
    return false;
  }

  /// Returns the `element` at a specified [index].
  ///
  /// If the index is invalid `null` will be returned.
  ///
  /// Example:
  /// ```dart
  /// final tuple = Tuple.fromList([2, 4, 6, 'a']);
  /// final element = tuple.elementAt(3);
  /// print(element); // a
  /// ```
  E? elementAt(int index) => _isValidIndex(index) ? _tupleList[index] : null;

  // helper function to validate index
  bool _isValidIndex(index) =>
      index >= 0 && index < _tupleList.length ? true : false;

  /// Returns the first `element` that satisfies a condition received
  /// as argument.
  ///
  /// If no `element` is found `null` will be returned instead.
  ///
  /// Example:
  /// ```dart
  /// final tuple = Tuple.fromList([2, 4, 6, 15, 18 'a']);
  /// print(tuple.firstWhere((element) => element is num && element > 10)); // 15
  /// ```
  firstWhere(bool Function(E element) test) {
    for (E element in _tupleList) {
      if (test(element)) return element;
    }
    return null;
  }

  /// Returns the last `element` that satisfies a condition received
  /// as argument.
  ///
  /// If no `element` is found `null` will be returned instead.
  ///
  /// Example:
  /// ```dart
  /// final tuple = Tuple.fromList([2, 4, 6, 15, 18 'a']);
  /// print(tuple.lastWhere((element) => element is num && element > 10)); // 18
  /// ```
  lastWhere(bool Function(E element) test) {
    for (E element in _tupleList.reversed) {
      if (test(element)) return element;
    }
    return null;
  }

  /// Returns the `index` of the first `element` that satisfies
  /// a condition received as argument.
  ///
  /// If no element is found `null` will be returned instead.
  ///
  /// Example:
  /// ```dart
  /// final tuple = Tuple.fromList([2, 4, 6, 15, 18 'a']);
  /// print(tuple.indexWhere((element) => element is num && element > 10)); // 3
  /// ```
  indexWhere(bool Function(E element) test) {
    for (E element in _tupleList) {
      if (test(element)) return _tupleList.indexOf(element);
    }
    return null;
  }

  @override
  String toString() {
    return _tupleList.toString();
  }
}


example/example.dart
import 'package:collection_extension/collection_extension.dart';

void main() {
  final stack = Stack();
  stack.push(3);
  stack.pushAll([5, 7, 9, 'a', 'b', 'c']);
  print(stack); // [3, 5, 7, 9, a, b, c]
  stack.pop(); // [3, 5, 7, 9, a, b]
  print(stack.contains('c')); // false
  stack.executeStack(); // 3 5 7 9 a b

  final tuple = Tuple.fromList([5, 7, 9, 'a', 'b', 'c', 13, 18]);
  print(tuple); // [5, 7, 9, a, b, c, 13, 18]
  print(tuple.contains('c')); // false
  print(tuple.elementAt(3)); // a
  print(tuple.firstWhere(
      (element) => element is num && element > 10 && element < 20)); // 13

  final specialQueue = SpecialQueue();
  specialQueue.push(7);
  specialQueue.pushAll([9, 3, 12, 'a', 'u', 19]);
  print(specialQueue); // [7, 9, 3, 12, a, u, 19]
  specialQueue.pop(); // [9, 3, 12, a, u, 19]
  final isRemoved = specialQueue.remove('a'); // [9, 3, 12, u, 19]
  print(isRemoved); //true
  final elementRemoved = specialQueue.removeFirst(); // [9, 3, 12, u, 19]
  print(elementRemoved); // 9
  specialQueue.executeQueue(); // 3 12 u 19
}


lib/src/special_queue.dart
/// A [SpecialQueue] is a `collection` that can be manipulated at both ends.
/// One can execute the specialQueue using [executeQueue] method.
///
/// The [SpecialQueue] can be created either empty or with `elements`
/// given through a [List].
///
/// For simplicity reasons [SpecialQueue] is a [List] with it's own
/// specific methods.
///
/// Example:
/// final specialQueue = SpecialQueue();
/// print(specialQueue.runtimeType); // SpecialQueue<Object>
///
/// // Adding elements to the specialQueue
/// specialQueue.push(5);
/// specialQueue.pushAll([3, 'a', 66]);
/// specialQueue.pushFirst(777);
/// print(specialQueue); // [777, 5, 3, 'a', 66]
///
/// // Removing elements from the specialQueue
/// specialQueue.removeFirst();
/// specialQueue.removeLast();
/// print(specialQueue); // [5, 3, 'a']
/// ```
class SpecialQueue<E> {
  /// The [List] used to store the [SpecialQueue] and its `elements`.
  final List _queueList;

  /// Whether this `collection` has no elements.
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue();
  /// print(specialQueue.isEmpty); // true
  /// ```
  bool get isEmpty => _queueList.isEmpty;

  /// Whether this `collection` has elements to pop.
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue();
  /// print(specialQueue.canPop); // false
  /// ```
  bool get canPop => _queueList.isNotEmpty;

  /// The number of `elements` in this `collection`.
  ///
  /// The valid indices are `0` through `length - 1`
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([2, 5, 9, 44]);
  /// print(specialQueue.length); // 4
  /// ```
  int get length => _queueList.length;

  /// The first element of the `collection`.
  ///
  /// If the `collection` is empty `null` will be return instead.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([2, 5, 9, 44]);
  /// print(specialQueue.peak) // 2
  /// ```
  E? get peak => _queueList.isNotEmpty ? _queueList[0] : null;

  /// Creates an empty [SpecialQueue] object.
  ///
  /// The `constructor` will initialise the [List] that manages
  /// [SpecialQueue] object.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue();
  /// print(specialQueue); // []
  /// ```
  SpecialQueue() : _queueList = <E>[];

  /// Creates a [SpecialQueue] object containing the elements of
  /// the [List] passed as argument.
  ///
  /// The [List] that manages [SpecialQueue] will be initialised with
  /// the elements of the [List] passed as argument.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 30, 'w']);
  /// print(specialQueue); // [5, 9, 30, w]
  /// ```
  SpecialQueue.fromList(List<E> list) : _queueList = list;

  /// Inserts [element] to the end of the [SpecialQueue].
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([7, 8]);
  /// specialQueue.push(5);
  /// print(specialQueue); // [7, 8, 5]
  /// ```
  void push(E element) => _queueList.add(element);

  /// Inserts all elements from [iterable] to the end of [SpecialQueue].
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([2, 4]);
  /// specialQueue.pushAll([5, 9, 77, 'a']);
  /// print(specialQueue); // [2, 4, 5, 9, 77, 'a']
  /// ```
  void pushAll(Iterable<E> iterable) => _queueList.addAll(iterable);

  /// Inserts [element] at the beginning of the [SpecialQueue].
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([2, 4]);
  /// specialQueue.pushFirst(9);
  /// print(specialQueue); // [9, 2, 4]
  /// ```
  void pushFirst(E element) => _queueList.insert(0, element);

  /// Removes the last `element` of the [SpecialQueue].
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// specialQueue.pop();
  /// print(specialQueue); // [5, 9]
  /// ```
  void pop() {
    if (canPop) _queueList.removeAt(0);
  }

  /// Checks whether [SpecialQueue] contains [element] or not.
  ///
  /// A `bool` object will be returned.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// bool hasElement = specialQueue.contains(9);
  /// print(hasElement); // true
  /// ```
  bool contains(E element) {
    for (E item in _queueList) {
      if (item == element) return true;
    }
    return false;
  }

  /// Removes [element] from the [SpecialQueue].
  ///
  /// If the list has `elements` it will try to find [element]
  /// and delete it.
  ///
  /// A `bool` object will be returned.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// bool isRemoved = specialQueue.remove(9);
  /// print(isRemoved); // true
  /// ```
  bool remove(E element) {
    if (canPop) {
      for (int i = 0; i < _queueList.length; i++) {
        if (element == _queueList[i]) {
          for (int j = i; j < _queueList.length - 1; j++) {
            _queueList[j] = _queueList[j + 1];
          }
          _queueList.removeLast();
          return true;
        }
      }
    }
    return false;
  }

  /// Removes the last `element` of the [SpecialQueue].
  ///
  /// The removed `element` will be returned.
  /// If the list has no `elements` it will return `null`.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// bool elementRemoved = specialQueue.removeLast();
  /// print(elementRemoved); // 'a'
  /// ```
  E? removeLast() {
    if (canPop) {
      final int indexToRemove = _queueList.length - 1;
      final removedItem = _queueList[indexToRemove];
      _queueList.removeAt(indexToRemove);
      return removedItem;
    }
    return null;
  }

  /// Removes the first `element` of the [SpecialQueue].
  ///
  /// The removed `element` will be returned.
  /// If the list has no `elements` it will return `null`.
  ///
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// bool elementRemoved = specialQueue.removeFirst();
  /// print(elementRemoved); // 5
  /// ```
  E? removeFirst() {
    if (canPop) {
      final int indexToRemove = 0;
      final removedItem = _queueList[indexToRemove];
      _queueList.removeAt(indexToRemove);
      return removedItem;
    }
    return null;
  }

  /// The [SpecialQueue] will be executed.
  ///
  /// Every `element` will be printed in order and removed from
  /// the [SpecialQueue].
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// specialQueue.executeQueue(); // 5 9 a
  /// print(specialQueue); // []
  /// ```
  void executeQueue() {
    for (E element in _queueList) {
      print(element);
    }
    popAll();
  }

  /// Removes all `elements` of the [SpecialQueue].
  ///
  /// Example:
  /// ```dart
  /// final specialQueue = SpecialQueue.fromList([5, 9, 'a']);
  /// specialQueue.popAll();
  /// print(specialQueue); // []
  /// ```
  void popAll() {
    if (canPop) _queueList.removeRange(0, _queueList.length);
  }

  @override
  String toString() {
    return _queueList.toString();
  }
}


example/new_package_example.dart
import 'dart:collection';

import 'package:new_package/new_package.dart';

void main() {
  // final manu = Stack();
  // manu.push(3);
  // manu.push(9);
  // manu.push('a');
  // manu.popAll();
  // manu.pushAll([25, 35, 99, 199, 399, 9999999]);
  // print(manu.canPop);
  // print(manu);
  // manu.executeStack();
  // print(manu);
  // manu.pop();
  // final boss = Stack.from([22, 33, 77, 100]);
  // print(boss);

  // final specialQueue = SpecialQueue();
  // specialQueue.executeQueue();
  // print(specialQueue.peak);
  // specialQueue.removeFirst();
  // specialQueue.push(5);
  // specialQueue.push(33);
  // specialQueue.push(44);
  // specialQueue.push(57);
  // specialQueue.push(570);
  // specialQueue.pushAll([20, 50, 34, 127]);
  // specialQueue.executeQueue();
  // print(specialQueue);
  // final specialQueue2 =
  //     SpecialQueue.fromList([20, 30, 40, '2', 60, 70, 80, 'a', 'c']);
  // // boss.pushLast(3);
  // final ceva = specialQueue2.remove(20);
  // print(ceva);
  // print(specialQueue2.runtimeType);

  final manu = Tuple.fromList(
      [3, 5, 7, 9, 11, 15, 19, 18, 22, 44, 55, 'a', 'b', 'c', 'd']);
  print(manu.contains('a'));
  print(manu.contains(5));
  print(manu.contains('5'));
  print(manu
      .firstWhere((element) => element is num && element > 10 && element < 20));
  print(manu.indexWhere((element) {
    return element == 2;
  }));
  print(manu
      .lastWhere((element) => element is num && element > 10 && element < 20));
  // print(manu.elementAt(15));
  // print(manu);
  print(manu.runtimeType);
}


example/collection_extension_example.dart
import 'dart:collection';

import 'package:collection_extension/collection_example.dart';

void main() {
  // final manu = Stack();
  // manu.push(3);
  // manu.push(9);
  // manu.push('a');
  // manu.popAll();
  // manu.pushAll([25, 35, 99, 199, 399, 9999999]);
  // print(manu.canPop);
  // print(manu);
  // manu.executeStack();
  // print(manu);
  // manu.pop();
  // final boss = Stack.from([22, 33, 77, 100]);
  // print(boss);

  // final specialQueue = SpecialQueue();
  // specialQueue.executeQueue();
  // print(specialQueue.peak);
  // specialQueue.removeFirst();
  // specialQueue.push(5);
  // specialQueue.push(33);
  // specialQueue.push(44);
  // specialQueue.push(57);
  // specialQueue.push(570);
  // specialQueue.pushAll([20, 50, 34, 127]);
  // specialQueue.executeQueue();
  // print(specialQueue);
  // final specialQueue2 =
  //     SpecialQueue.fromList([20, 30, 40, '2', 60, 70, 80, 'a', 'c']);
  // // boss.pushLast(3);
  // final ceva = specialQueue2.remove(20);
  // print(ceva);
  // print(specialQueue2.runtimeType);

  final manu = Tuple.fromList(
      [3, 5, 7, 9, 11, 15, 19, 18, 22, 44, 55, 'a', 'b', 'c', 'd']);
  print(manu.contains('a'));
  print(manu.contains(5));
  print(manu.contains('5'));
  print(manu
      .firstWhere((element) => element is num && element > 10 && element < 20));
  print(manu.indexWhere((element) {
    return element == 2;
  }));
  print(manu
      .lastWhere((element) => element is num && element > 10 && element < 20));
  // print(manu.elementAt(15));
  // print(manu);
  print(manu.runtimeType);
}
